server/src/test/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainActionTests.java

```diff
@@ -109,6 +109,20 @@ public ShardAllocationDecision decideShardAllocation(ShardRouting shard, Routing
             cae.getCurrentNode().getId(),
             cae.getCurrentNode().getName(),
             cae.getCurrentNode().getAddress(),
+            "data",  // begin roles array
+            "data_cold",
+            "data_content",
+            "data_frozen",
+            "data_hot",
+            "data_warm",
+            "index",
+            "ingest",
+            "master",
+            "ml",
+            "remote_cluster_client",
+            "search",
+            "transform",
+            "voting_only",  // end roles array
```

Comment:
This would need updating if we ever add a new node role, but I don't think the test really cares about the exact roles. How about constructing the expected string dynamically:

```suggestion
            cae.getCurrentNode().getRoles().stream().map(r -> '"' + r.roleName() + '"').collect(Collectors.joining(", ", "[", "]")),
```

(and then the format string just needs `"roles": %s`)

Reply:
Sounds great, figuring out the formatting to get everything to work in Java was kicking my ass :)